.github/workflows/deploy.yml

```diff
@@ -0,0 +1,99 @@
+name: Deploy
+
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+    tags:
+      # pre-release tag
+      - "202[3-9].[0-9][0-9].[0-9]+-rc[0-9]+"
+      # release tags
+      - "202[3-9].[0-9][0-9].[0-9]+"
+
+defaults:
+  run:
+    shell: bash
+
+jobs:
+  tests-ui:
+    uses: ./.github/workflows/tests-ui.yml
+    if: github.ref_type == 'tag'
+
+  tests-pytest:
+    uses: ./.github/workflows/tests-pytest.yml
+    if: github.ref_type == 'tag'
+
+  deploy:
+    runs-on: ubuntu-latest
+    needs: [tests-ui, tests-pytest]
+    if: (!cancelled())
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - uses: actions/setup-python@v5
+        with:
+          python-version-file: .github/workflows/.python-version
+          cache: pip
+          cache-dependency-path: "**/pyproject.toml"
+
+      - name: Write python packages to file
+        run: |
+          python -m venv .venv
+          source .venv/bin/activate
+          pip install pipdeptree
+          pip install -e .
+          pipdeptree
+          pipdeptree >> pems/static/requirements.txt
+
+      - name: Write commit SHA to file
+        run: echo "${{ github.sha }}" >> pems/static/sha.txt
+
+      - name: Docker Login to GitHub Container Registry
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.actor }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v3
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v3
+
+      - name: Build, tag, and push image to GitHub Container Registry
+        uses: docker/build-push-action@v6
+        with:
+          platforms: linux/amd64,linux/arm64
+          builder: ${{ steps.buildx.outputs.name }}
+          build-args: GIT-SHA=${{ github.sha }}
+          cache-from: type=gha,scope=compilerla
+          cache-to: type=gha,scope=compilerla,mode=max
+          context: .
+          file: appcontainer/Dockerfile
+          push: true
+          tags: ghcr.io/${{ github.repository }}:${{ github.sha }}
```

Comment:
Can we try and tag with the git tag as well?

Reply:
Yep, since it looks like [`github.ref_name`](https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/accessing-contextual-information-about-workflow-runs#github-context) is what we need, would this tag format (`sha-version`) work for us?

```
tags: ghcr.io/${{ github.repository }}:${{ github.sha }}-${{ github.ref_name }}
```

Reply:
I was actually suggesting the git tag _in addition to_ the SHA (i.e. keeping them distinct). I think you can do:

```yml
tags: |
  ghcr.io/${{ github.repository }}:${{ github.ref_name }}
  ghcr.io/${{ github.repository }}:${{ github.sha }}
```

Reply:
Check out [these docs referencing the `docker/metadata-action`](https://docs.docker.com/build/ci/github-actions/manage-tags-labels/) -- that seems useful?

Reply:
Got it, thanks!